File: liaisons/kotlin/src/main/kotlin/io/github/airdesktopproject/asl/Abi.kt
// L'ABI C, transcrite pour l'API FFM. **Rien d'idiomatique ici.**
//
// Ce fichier est la copie mot pour mot de `crates/asl-client-ffi/include/asl.h`.
// Il ne traduit pas les fautes, ne nomme rien autrement, n'enveloppe rien :
// c'est le rôle de `Asl.kt`. Les séparer rend visible la seule question qui
// compte ici — **est-ce que ce fichier dit la même chose que l'en-tête ?** —, et
// `essais/Essais.kt` y répond en lisant les deux.
//
// ── POURQUOI `java.lang.foreign` ET NON JNI, JNA OU JNR ────────────────────
//
// **JNI demande d'écrire du C.** Une glu native à compiler, à distribuer par
// plate-forme, et à maintenir — c'est-à-dire exactement ce que la contrainte C4
// de ce dépôt refuse partout ailleurs.
//
// **JNA et JNR sont des dépendances**, et JNA embarque ses propres stubs natifs.
// Ce que cette bibliothèque tire, ses porteurs l'installent.
//
// L'API FFM est dans le JDK depuis la version 22, sans dépendance et sans une
// ligne de C. C'est la même décision que `ctypes` en Python et `fiddle` en Ruby,
// prise pour la même raison.
//
// **CE QU'ELLE COÛTE, ET QUI EST RÉEL** : elle exige un JDK 22 ou plus, et
// `--enable-native-access` au lancement. Android n'en dispose pas — voir
// `liaisons/kotlin/README.md`.

package io.github.airdesktopproject.asl

import java.lang.foreign.Arena
import java.lang.foreign.FunctionDescriptor
import java.lang.foreign.Linker
import java.lang.foreign.MemoryLayout
import java.lang.foreign.MemoryLayout.PathElement
import java.lang.foreign.MemorySegment
import java.lang.foreign.SymbolLookup
import java.lang.foreign.ValueLayout
import java.lang.invoke.MethodHandle
import java.nio.file.Files
import java.nio.file.Path

/** L'objet natif n'a pas été trouvé.
 *
 * **CE N'EST PAS UNE FAUTE D'EXÉCUTION, C'EST UNE INSTALLATION INCOMPLÈTE**, et
 * le message le dit — un `UnsatisfiedLinkError` nu enverrait chercher une panne
 * là où il manque un fichier.
 */
public class BibliothequeIntrouvable(message: String) : RuntimeException(message)

/** La transcription de `asl.h`. */
internal object Abi {
    // ── LES CODES ───────────────────────────────────────────────────────────
    const val OK: Int = 0
    const val ARGUMENT: Int = -1
    const val CONFIGURATION: Int = -2
    const val INJOIGNABLE: Int = -3
    const val REFUSE: Int = -4
    const val TAMPON_TROP_PETIT: Int = -5
    const val INTERNE: Int = -6
    const val PAS_D_IDENTITE: Int = -7
    const val DEJA: Int = -8

    const val IDENTIFIANT_OCTETS: Int = 29
    const val GRAINE_OCTETS: Int = 32

    const val TCP: Int = 1
    const val UDP: Int = 2

    const val REFLEXIF: Int = 1
    const val ANNONCE: Int = 2

    const val JOIGNABLE: Int = 1
    const val INJOIGNABLE_POINT: Int = 2
    const val NON_SONDE: Int = 3
    const val EN_COURS: Int = 4

    // ── LES DISPOSITIONS ────────────────────────────────────────────────────
    //
    // **AUCUN `paddingLayout` N'EST NÉCESSAIRE**, et c'est un fait à vérifier, pas
    // à supposer : `MemoryLayout.structLayout` REFUSE une disposition dont un
    // membre serait mal aligné. Si l'en-tête introduisait un jour un trou
    // implicite, cette classe ne se chargerait pas — ce qui est le bon moment
    // pour l'apprendre.

    val POINT: MemoryLayout = MemoryLayout.structLayout(
        ValueLayout.JAVA_SHORT.withName("port"),
        ValueLayout.JAVA_BYTE.withName("protocole"),
        ValueLayout.JAVA_BYTE.withName("reserve"),
    )

    val CANDIDAT: MemoryLayout = MemoryLayout.structLayout(
        MemoryLayout.sequenceLayout(16, ValueLayout.JAVA_BYTE).withName("adresse"),
        ValueLayout.JAVA_SHORT.withName("port"),
        ValueLayout.JAVA_BYTE.withName("protocole"),
        ValueLayout.JAVA_BYTE.withName("famille"),
        ValueLayout.JAVA_BYTE.withName("origine"),
        ValueLayout.JAVA_BYTE.withName("verdict"),
        MemoryLayout.sequenceLayout(2, ValueLayout.JAVA_BYTE).withName("reserve"),
    )

    val ETAT: MemoryLayout = MemoryLayout.structLayout(
        ValueLayout.JAVA_LONG.withName("attaches"),
        ValueLayout.JAVA_LONG.withName("ruptures"),
        ValueLayout.JAVA_BYTE.withName("attachee"),
        ValueLayout.JAVA_BYTE.withName("abandonnee"),
        MemoryLayout.sequenceLayout(6, ValueLayout.JAVA_BYTE).withName("reserve"),
    )

    /** Le décalage d'un champ, par son nom. */
    fun decalage(disposition: MemoryLayout, champ: String): Long =
        disposition.byteOffset(PathElement.groupElement(champ))

    // **LES TAILLES SONT VÉRIFIÉES AU CHARGEMENT DE LA CLASSE.**
    //
    // C'est le pendant JVM des `static_assert` du côté C++ et des `const _: () =
    // assert!` du côté Rust. Si cette transcription compte mal, il vaut mieux
    // qu'elle refuse de se charger que d'écrire un port dans un champ de
    // protocole pendant six mois.
    init {
        require(POINT.byteSize() == 4L) { "asl_point fait ${POINT.byteSize()} octets, pas 4" }
        require(CANDIDAT.byteSize() == 24L) {
            "asl_candidat fait ${CANDIDAT.byteSize()} octets, pas 24"
        }
        require(ETAT.byteSize() == 24L) { "asl_etat_t fait ${ETAT.byteSize()} octets, pas 24" }
    }

    // ── LES TYPES DE L'ABI, TELS QUE LA PLATE-FORME LES DÉFINIT ─────────────
    //
    // **`size_t` N'EST PAS `long`.** Il l'est sur les plates-formes qui nous
    // intéressent aujourd'hui, et il ne l'était pas hier. L'API FFM publie les
    // dispositions canoniques de la plate-forme courante : les employer est ce
    // qui évite d'écrire `JAVA_LONG` et de découvrir le contraire ailleurs.
    private val LIEUR: Linker = Linker.nativeLinker()
    val TAILLE: ValueLayout = LIEUR.canonicalLayouts()["size_t"] as ValueLayout
    private val ENTIER = ValueLayout.JAVA_INT
    private val ADRESSE = ValueLayout.ADDRESS

    /** Les onze fonctions, avec leur signature. */
    private val SIGNATURES: Map<String, FunctionDescriptor> = mapOf(
        "asl_version" to FunctionDescriptor.ofVoid(ADRESSE, ADRESSE, ADRESSE),
        "asl_faute_texte" to FunctionDescriptor.of(ADRESSE, ENTIER),
        "asl_client_neuf" to FunctionDescriptor.of(ENTIER, ADRESSE),
        "asl_client_annuaire" to FunctionDescriptor.of(ENTIER, ADRESSE, ADRESSE, ADRESSE),
        "asl_client_racines" to FunctionDescriptor.of(ENTIER, ADRESSE, ADRESSE, TAILLE),
        "asl_client_identite" to FunctionDescriptor.of(ENTIER, ADRESSE, ADRESSE, ADRESSE),
        "asl_client_libere" to FunctionDescriptor.ofVoid(ADRESSE),
        "asl_enroler" to FunctionDescriptor.of(ENTIER, ADRESSE, ADRESSE, ADRESSE, ADRESSE),
        "asl_annoncer" to FunctionDescriptor.of(ENTIER, ADRESSE, ADRESSE, ADRESSE, TAILLE),
        "asl_etat" to FunctionDescriptor.of(ENTIER, ADRESSE, ADRESSE),
        "asl_ou" to FunctionDescriptor.of(
            ENTIER, ADRESSE, ADRESSE, ADRESSE, ADRESSE, TAILLE, ADRESSE,
        ),
    )

    /** Comment l'objet natif s'appelle, selon le système. */
    fun nomsPossibles(): List<String> {
        val systeme = System.getProperty("os.name").orEmpty().lowercase()
        return when {
            systeme.contains("mac") -> listOf("libasl_client_ffi.dylib")
            systeme.contains("win") -> listOf("asl_client_ffi.dll")
            else -> listOf("libasl_client_ffi.so")
        }
    }

    /** Le couple système-architecture, tel qu'il nomme un répertoire de
     * ressources.
     *
     * **`os.arch` NE DIT PAS LA MÊME CHOSE QUE `rustc`** : la JVM dit `amd64` là
     * où Rust dit `x86_64`, et `arm64` là où il dit `aarch64`. Les normaliser ici
     * est ce qui évite de chercher un répertoire qui n'existera jamais.
     */
    fun cleDePlateforme(): String {
        val systeme = System.getProperty("os.name").orEmpty().lowercase()
        val machine = System.getProperty("os.arch").orEmpty().lowercase()
        val quel = when {
            systeme.contains("mac") -> "macos"
            systeme.contains("win") -> "windows"
            else -> "linux"
        }
        val laquelle = when (machine) {
            "amd64", "x86_64" -> "x86_64"
            "aarch64", "arm64" -> "aarch64"
            else -> machine
        }
        return "$quel-$laquelle"
    }

    /** Extrait l'objet natif du JAR, s'il s'y trouve.
     *
     * # POURQUOI IL FAUT EXTRAIRE, ET QUE C'EST LE SEUL CAS DES CINQ
     *
     * Python et Ruby posent l'objet À CÔTÉ de leur paquet, et le trouvent par un
     * chemin. **Un JAR n'est pas un système de fichiers** : une ressource y est
     * un flux, et `SymbolLookup.libraryLookup` veut un chemin. Il n'y a donc pas
     * d'autre façon que de l'écrire quelque part.
     *
     * Le fichier est temporaire et marqué pour effacement à la sortie de la JVM.
     * **Le coût est réel** : un mégaoctet écrit et une ouverture de fichier au
     * premier appel, une fois par processus. C'est le prix d'un JAR autonome, et
     * un porteur qui ne le veut pas pose `ASL_BIBLIOTHEQUE`.
     */
    private fun depuisLesRessources(): Path? {
        val nom = nomsPossibles().first()
        val chemin = "/natif/${cleDePlateforme()}/$nom"
        val flux = Abi::class.java.getResourceAsStream(chemin) ?: return null
        return flux.use { entrant ->
            val vers = Files.createTempFile("asl-", "-$nom")
            vers.toFile().deleteOnExit()
            Files.copy(entrant, vers, java.nio.file.StandardCopyOption.REPLACE_EXISTING)
            vers
        }
    }

    /** Où chercher, dans l'ordre.
     *
     * **`ASL_BIBLIOTHEQUE` PASSE AVANT TOUT.** C'est ce qui permet d'éprouver
     * cette liaison contre une construction locale sans l'installer, et à un
     * porteur de pointer l'objet qu'il a compilé pour son architecture.
     *
     * Ensuite `java.library.path`, que la JVM offre déjà pour cela. Puis le JAR
     * lui-même — voir [depuisLesRessources].
     */
    fun cheminsCandidats(): List<Path> {
        val chemins = mutableListOf<Path>()
        System.getenv("ASL_BIBLIOTHEQUE")?.takeIf { it.isNotEmpty() }?.let {
            chemins.add(Path.of(it))
        }
        val repertoires = System.getProperty("java.library.path").orEmpty()
            .split(java.io.File.pathSeparator)
            .filter { it.isNotEmpty() }
        for (repertoire in repertoires) {
            for (nom in nomsPossibles()) {
                chemins.add(Path.of(repertoire, nom))
            }
        }
        depuisLesRessources()?.let { chemins.add(it) }
        return chemins
    }

    /** Les fonctions, une fois la bibliothèque chargée. */
    class Fonctions internal constructor(private val table: Map<String, MethodHandle>) {
        operator fun get(nom: String): MethodHandle =
            table[nom] ?: error("`$nom` n'a pas été lié : ce n'est pas censé arriver")
    }

    /** Charge la bibliothèque native et lie ses onze fonctions.
     *
     * L'`Arena.global()` n'est pas un oubli : cette bibliothèque vit aussi
     * longtemps que la JVM. La décharger demanderait de garantir qu'aucun fil
     * natif ne tourne encore — or c'est précisément ce qu'une annonce tenue fait.
     */
    fun charger(chemin: Path? = null): Fonctions {
        val essayes = mutableListOf<String>()
        val candidats = chemin?.let { listOf(it) } ?: cheminsCandidats()

        var recherche: SymbolLookup? = null
        for (candidat in candidats) {
            essayes.add(candidat.toString())
            if (Files.exists(candidat)) {
                recherche = SymbolLookup.libraryLookup(candidat, Arena.global())
                break
            }
        }

        if (recherche == null) {
            // En dernier, le chargeur du système.
            for (nom in nomsPossibles()) {
                essayes.add(nom)
                try {
                    recherche = SymbolLookup.libraryLookup(nom, Arena.global())
                    break
                } catch (_: IllegalArgumentException) {
                    // Ce nom-là n'existe pas ; on essaie le suivant.
                }
            }
        }

        if (recherche == null) {
            throw BibliothequeIntrouvable(
                "l'objet natif d'asl est introuvable.\n" +
                    "Cherché : ${essayes.joinToString(", ")}\n" +
                    "Construisez-le avec `cargo build --release` dans le dépôt, puis " +
                    "posez ASL_BIBLIOTHEQUE sur target/release/${nomsPossibles().first()}",
            )
        }

        val table = SIGNATURES.entries.associate { (nom, signature) ->
            val adresse = recherche.find(nom).orElseThrow {
                BibliothequeIntrouvable(
                    "la bibliothèque native n'exporte pas `$nom`. " +
                        "Elle ne correspond pas à cette liaison.",
                )
            }
            nom to LIEUR.downcallHandle(adresse, signature)
        }
        return Fonctions(table)
    }
}

File: liaisons/ruby/lib/asl/abi.rb
# frozen_string_literal: true

# L'ABI C, transcrite pour `fiddle`. **Rien d'idiomatique ici.**
#
# Ce fichier est la copie mot pour mot de `crates/asl-client-ffi/include/asl.h`.
# Il ne traduit pas les erreurs, ne nomme rien autrement, n'enveloppe rien :
# c'est le rôle de `lib/asl.rb`. Les séparer rend visible la seule question qui
# compte ici — **est-ce que ce fichier dit la même chose que l'en-tête ?** —, et
# `essais/test_abi.rb` y répond en lisant les deux.
#
# ── POURQUOI `fiddle` ET NON LA GEMME `ffi` ─────────────────────────────────
#
# La gemme `ffi` est plus agréable. Elle est aussi une **extension native** :
# l'installer compile du C dans l'environnement de qui nous embarque. Ce dépôt a
# une contrainte entière (C4) qui tient à ce qu'aucun C n'entre — la refuser ici
# et l'accepter par la porte du gestionnaire de gemmes n'aurait aucun sens.
#
# `fiddle` est dans la distribution de Ruby, et ne coûte rien à personne.
#
# ── POURQUOI `pack` ET `unpack`, ET NON UN CONSTRUCTEUR DE STRUCTURES ───────
#
# `pack` n'insère JAMAIS de bourrage. Ce serait normalement disqualifiant pour
# lire une structure C — sauf que celles-ci n'en ont aucun : l'en-tête range ses
# champs du plus large au plus étroit, exprès, pour que cinq langages qui
# calculent la disposition chacun de son côté tombent d'accord.
#
# **La disposition packée EST donc la disposition C**, et c'est le bénéfice
# direct de cette décision-là. Une structure à trou aurait rendu ceci impossible.
#
# Les gabarits emploient les tailles NATIVES (`S`, `Q`) et non des variantes
# petit-boutistes : la structure est native, pas un format de fil.

require "fiddle"

module Asl
  # La transcription de `asl.h`.
  module Abi
    # ── LES CODES ───────────────────────────────────────────────────────────
    OK = 0
    ARGUMENT = -1
    CONFIGURATION = -2
    INJOIGNABLE = -3
    REFUSE = -4
    TAMPON_TROP_PETIT = -5
    INTERNE = -6
    PAS_D_IDENTITE = -7
    DEJA = -8

    IDENTIFIANT_OCTETS = 29
    GRAINE_OCTETS = 32

    TCP = 1
    UDP = 2

    REFLEXIF = 1
    ANNONCE = 2

    JOIGNABLE = 1
    INJOIGNABLE_POINT = 2
    NON_SONDE = 3
    EN_COURS = 4

    # ── LES DISPOSITIONS ────────────────────────────────────────────────────

    # `asl_point` : `uint16 port`, `uint8 protocole`, `uint8 reserve`.
    GABARIT_POINT = "SCC"
    TAILLE_POINT = 4

    # `asl_candidat` : 16 octets d'adresse, puis port, protocole, famille,
    # origine, verdict, et deux octets réservés.
    GABARIT_CANDIDAT = "a16SCCCCa2"
    TAILLE_CANDIDAT = 24

    # `asl_etat_t` : deux compteurs de 64 bits, deux drapeaux, six réservés.
    GABARIT_ETAT = "QQCCa6"
    TAILLE_ETAT = 24

    # **LES NOMS DES CHAMPS, DANS L'ORDRE.**
    #
    # `unpack` rend un tableau : les noms n'existent nulle part, et un champ
    # déplacé ne changerait NI la taille NI le comportement de `pack` — on lirait
    # simplement un port là où il y a un protocole, sans que rien ne proteste.
    #
    # Ces listes sont ce que `essais/test_abi.rb` compare aux champs déclarés dans
    # l'en-tête. Elles ne servent qu'à cela, et c'est suffisant pour exister.
    CHAMPS_POINT = %i[port protocole reserve].freeze
    CHAMPS_CANDIDAT = %i[adresse port protocole famille origine verdict reserve].freeze
    CHAMPS_ETAT = %i[attaches ruptures attachee abandonnee reserve].freeze

    # **LES TAILLES SONT VÉRIFIÉES AU CHARGEMENT, ET NON DANS UN ESSAI.**
    #
    # C'est le pendant Ruby des `const _: () = assert!` du côté Rust. Si ce
    # fichier compte mal, il vaut mieux qu'il refuse de se charger que d'écrire un
    # port dans un champ de protocole pendant six mois.
    # Combien d'octets ce gabarit occupe, mesuré en l'employant.
    #
    # Il n'y a pas de « taille d'un gabarit » en Ruby : on en emballe un
    # exemplaire vide et l'on regarde. `a<N>` veut une chaîne, les autres
    # directives un entier — d'où ce petit tri.
    def self.mesurer(gabarit)
      exemplaire = gabarit.scan(/[a-zA-Z]\d*/).map do |directive|
        directive.start_with?("a") ? "" : 0
      end
      exemplaire.pack(gabarit).bytesize
    end

    {
      GABARIT_POINT => TAILLE_POINT,
      GABARIT_CANDIDAT => TAILLE_CANDIDAT,
      GABARIT_ETAT => TAILLE_ETAT
    }.each do |gabarit, taille|
      mesure = mesurer(gabarit)
      next if mesure == taille

      raise LoadError,
            "le gabarit #{gabarit.inspect} fait #{mesure} octets, et l'ABI en " \
            "annonce #{taille}. Cette liaison ne correspond pas à la " \
            "bibliothèque native : ne l'utilisez pas."
    end

    # ── LE CHARGEMENT ───────────────────────────────────────────────────────

    # L'objet natif n'a pas été trouvé.
    #
    # **CE N'EST PAS UNE FAUTE D'EXÉCUTION, C'EST UNE INSTALLATION INCOMPLÈTE**,
    # et le message le dit — l'exception brute de `fiddle` enverrait chercher une
    # panne là où il manque un fichier.
    class BibliothequeIntrouvable < StandardError; end

    # Comment l'objet natif s'appelle, selon le système.
    def self.noms_possibles
      case RbConfig::CONFIG["host_os"]
      when /darwin/ then ["libasl_client_ffi.dylib"]
      when /mswin|mingw/ then ["asl_client_ffi.dll"]
      else ["libasl_client_ffi.so"]
      end
    end

    # Où chercher, dans l'ordre.
    #
    # **`ASL_BIBLIOTHEQUE` PASSE AVANT TOUT.** C'est ce qui permet d'éprouver
    # cette liaison contre une construction locale sans l'installer, et à un
    # porteur de pointer l'objet qu'il a compilé pour son architecture.
    def self.chemins_candidats
      chemins = []
      impose = ENV["ASL_BIBLIOTHEQUE"]
      chemins << impose if impose && !impose.empty?
      ici = File.dirname(__dir__)
      noms_possibles.each { |nom| chemins << File.join(ici, nom) }
      chemins
    end

    # Les onze fonctions, une fois chargées.
    #
    # **CHAQUE SIGNATURE EST DÉCLARÉE À LA MAIN.** C'est le prix de `fiddle`, et
    # il est réel : une déclaration fausse ne se voit pas au chargement, elle
    # corrompt la pile à l'appel.
    SIGNATURES = {
      asl_version: [[Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP],
                    Fiddle::TYPE_VOID],
      asl_faute_texte: [[Fiddle::TYPE_INT32_T], Fiddle::TYPE_VOIDP],
      asl_client_neuf: [[Fiddle::TYPE_VOIDP], Fiddle::TYPE_INT32_T],
      asl_client_annuaire: [[Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP],
                           Fiddle::TYPE_INT32_T],
      asl_client_racines: [[Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP, Fiddle::TYPE_SIZE_T],
                          Fiddle::TYPE_INT32_T],
      asl_client_identite: [[Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP],
                           Fiddle::TYPE_INT32_T],
      asl_client_libere: [[Fiddle::TYPE_VOIDP], Fiddle::TYPE_VOID],
      asl_enroler: [[Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP,
                     Fiddle::TYPE_VOIDP], Fiddle::TYPE_INT32_T],
      asl_annoncer: [[Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP,
                      Fiddle::TYPE_SIZE_T], Fiddle::TYPE_INT32_T],
      asl_etat: [[Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP], Fiddle::TYPE_INT32_T],
      asl_ou: [[Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP,
                Fiddle::TYPE_VOIDP, Fiddle::TYPE_SIZE_T, Fiddle::TYPE_VOIDP],
               Fiddle::TYPE_INT32_T]
    }.freeze

    # Charge la bibliothèque native et rend ses fonctions, par nom.
    def self.charger(chemin = nil)
      essayes = []
      candidats = chemin ? [chemin.to_s] : chemins_candidats

      poignee = nil
      candidats.each do |candidat|
        essayes << candidat
        next unless File.exist?(candidat)

        poignee = Fiddle.dlopen(candidat)
        break
      end

      if poignee.nil?
        # En dernier, le chargeur du système — qui connaît `LD_LIBRARY_PATH`,
        # `/etc/ld.so.conf` et les répertoires d'installation.
        noms_possibles.each do |nom|
          essayes << nom
          begin
            poignee = Fiddle.dlopen(nom)
            break
          rescue Fiddle::DLError # rubocop:disable Lint/SuppressedException
          end
        end
      end

      if poignee.nil?
        raise BibliothequeIntrouvable,
              "l'objet natif d'asl est introuvable.\n" \
              "Cherché : #{essayes.join(', ')}\n" \
              "Construisez-le avec `cargo build --release` dans le dépôt, puis " \
              "posez ASL_BIBLIOTHEQUE sur target/release/#{noms_possibles.first}"
      end

      declarer(poignee)
    end

    # Pose les signatures. **`need_gvl: false` EST EXPLICITE, ET NON SUPPOSÉ.**
    #
    # Le GVL doit être RELÂCHÉ pendant l'appel : `asl_ou` attend jusqu'à vingt
    # secondes qu'un annuaire réponde, et `asl_client_libere` jusqu'à deux qu'une
    # annonce se retire proprement. Avec le GVL tenu, tout le processus hôte —
    # son serveur web, ses tâches, ses fils — gèlerait pendant ce temps.
    #
    # C'est la valeur par défaut de `fiddle` aujourd'hui. Elle est écrite quand
    # même : un défaut qui change ne casse rien à la compilation, il fige une
    # application en production.
    def self.declarer(poignee)
      SIGNATURES.each_with_object({}) do |(nom, (arguments, retour)), fonctions|
        fonctions[nom] = Fiddle::Function.new(
          poignee[nom.to_s], arguments, retour, name: nom.to_s, need_gvl: false
        )
      end
    end
  end
end
